### PR TITLE
CDAP-1724: Meta metrics for AggregatedMetricsCollectionService

### PR DESCRIPTION
**Pull request by jwang47**
http://builds.cask.co/browse/CDAP-DUT1322

Now we have

* total number of metric values processed
* processing delay time

```
$ curl -X POST http://localhost:10000/v3/metrics/query?context=component.metrics~processor&metric=system.processed.count | python -mjson.tool
{
    "endTime": 0,
    "series": [
        {
            "data": [
                {
                    "time": 0,
                    "value": 271
                }
            ],
            "grouping": {},
            "metricName": "system.metrics.global.processed.count"
        }
    ],
    "startTime": 0
}

$ curl -X POST http://localhost:10000/v3/metrics/query?context=component.metrics~processor&metric=system.processed.delay.avg | python -mjson.tool
{
    "endTime": 0,
    "series": [
        {
            "data": [
                {
                    "time": 0,
                    "value": 966
                }
            ],
            "grouping": {},
            "metricName": "system.metrics.global.processed.delay.ms"
        }
    ],
    "startTime": 0
}
```

### CODE REVIEW

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionService.java [hunk 1]
```diff
@@ -87,15 +88,25 @@ protected void publish(Iterator<MetricValue> metrics) throws Exception {
     while (metrics.hasNext()) {
       // Encode each MetricRecord into bytes and make it an individual kafka message in a message set.
       MetricValue value = metrics.next();
-      recordWriter.encode(value, encoder);
-      // partitioning by the context
-      preparer.add(ByteBuffer.wrap(encoderOutputStream.toByteArray()), getPartitionKey(value));
-      encoderOutputStream.reset();
+      metaMetricsComputer.visitMetric(value);
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionService.java]: so you are emitting latency metric for every record ?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java [hunk 2]
```diff
@@ -86,7 +91,33 @@ protected final void runOneIteration() throws Exception {
...
+          numMetrics++;
+
+          long timestampMs = TimeUnit.MILLISECONDS.convert(metricValue.getTimestamp(), TimeUnit.SECONDS);
+          long nowMs = System.currentTimeMillis();
+          long delayMs = nowMs - timestampMs;
+
+          MetricValue processDelayMetric = new MetricValue(
+            ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, "system"),
+            "metrics.global.processed.delay.ms", timestamp, delayMs, MetricType.GAUGE);
+          metaMetrics.add(processDelayMetric);
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: so if you call visitMetric on every record received at KafkaMetricsCollectionService, and there will be multiple records for the same timestamp, we will be calculating the latency for them and using Gauge will just override them. which seems incorrect. we should either sample or use counter. correct me if i am missing anything.

Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: also this line can potentially cause overflow.

File: cdap-watchdog/src/test/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionServiceTest.java [hunk 3]
```diff
@@ -194,9 +194,8 @@ public void finished() {
       }
     });
 
-    Assert.assertTrue(semaphore.tryAcquire(expected.size(), 5, TimeUnit.SECONDS));
+    Assert.assertTrue(semaphore.tryAcquire(expected.size(), 15, TimeUnit.SECONDS));
```
Comment (gsps1) [inline cdap-watchdog/src/test/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionServiceTest.java]: why 10 more seconds ? please add some comment

File: cdap-watchdog/src/test/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionServiceTest.java [hunk 4]
```diff
@@ -194,9 +194,8 @@ public void finished() {
...
 
-    Assert.assertEquals(expected.size(), metrics.size());
```
Comment (gsps1) [inline cdap-watchdog/src/test/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionServiceTest.java]: why ?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/MapReduceCounterCollectionService.java [hunk 5]
```diff
@@ -46,26 +46,35 @@ protected void startUp() throws Exception {
...
+    while (metaMetricsIterator.hasNext()) {
+      publishMetric(metaMetricsIterator.next());
+    }
+  }
 
-      // The format of the counters:
-      // * counter group name: "cdap.<tag_name>.<tag_value>[.<tag_name>.<tag_value>[...]]
-      // * counter name: metric name
+  private void publishMetric(MetricValue record) {
+    // The format of the counters:
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/MapReduceCounterCollectionService.java]: these are just formatting changes right ?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java (hunk 2 above)
Comment (jwang47) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: Yeah, we should sample. Why would this cause overflow though?

Comment (jwang47) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: Do you mean we may be storing too many `metaMetrics` in memory? Maybe we can have meta metrics per metric and per metric batch then?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java [hunk 6]
```diff
@@ -86,7 +91,33 @@ protected final void runOneIteration() throws Exception {
...
-      publish(metricsItor);
+      publish(metricsItor, new MetaMetricsComputer() {
+        private List<MetricValue> metaMetrics = Lists.newArrayList();
+        private int numMetrics = 0;
+
+        @Override
+        public void visitMetric(MetricValue metricValue) {
+          numMetrics++;
+
+          long timestampMs = TimeUnit.MILLISECONDS.convert(metricValue.getTimestamp(), TimeUnit.SECONDS);
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: I think using `TimeUnit.SECONDS.toMillis(metricValue.getTimestamp());` is simpler

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java [hunk 7]
```diff
@@ -86,7 +91,33 @@ protected final void runOneIteration() throws Exception {
...
+          long nowMs = System.currentTimeMillis();
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: Doing this for every record is not free. You only need to do this if the metric value timestamp changed (in 1 second, there can be a lot of metric values).

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java [hunk 8]
```diff
@@ -145,6 +176,8 @@ protected MetricValue computeNext() {
         return endOfData();
       }
     };
+
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: Why extra empty lines?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java [hunk 9]
```diff
@@ -75,9 +78,11 @@ public AggregatedMetricsEmitter load(EmitterKey key) throws Exception {
    * This method is guaranteed not to get concurrent calls.
    *
    * @param metrics collection of {@link co.cask.cdap.api.metrics.MetricValue} to publish.
+   * @param metaMetricsComputer computes metrics about metrics
    * @throws Exception if there is error raised during publish.
    */
-  protected abstract void publish(Iterator<MetricValue> metrics) throws Exception;
+  protected abstract void publish(Iterator<MetricValue> metrics,
+                                  MetaMetricsComputer metaMetricsComputer) throws Exception;
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: Why need to change this? Isn't it better to have the Iterator do it internally while iterating, rather than sub-class need to calls it explicitly?

Comment (jwang47) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: True, I'll have some Iterator that does this.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/MapReduceCounterCollectionService.java (hunk 5 above)
Comment (jwang47) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/MapReduceCounterCollectionService.java]: yeah

File: cdap-watchdog/src/test/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionServiceTest.java (hunk 4 above)
Comment (jwang47) [inline cdap-watchdog/src/test/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionServiceTest.java]: Still figuring out this test.. not able to run due to missing some native libs on my machine.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java [hunk 10]
```diff
@@ -46,13 +49,17 @@
...
 
   private final LoadingCache<Map<String, String>, MetricsCollector> collectors;
   private final LoadingCache<EmitterKey, AggregatedMetricsEmitter> emitters;
+  private final boolean recordMetaMetrics;
 
   private ScheduledExecutorService executorService;
+  private long timestampLastSampled;
 
-  public AggregatedMetricsCollectionService() {
+  public AggregatedMetricsCollectionService(boolean recordMetaMetrics) {
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: may be its better to have  ``recordMetaMetrics``` to be true by default and have a disableMetaMetrics method, so the subclasses can call them if needed. so you don't have to change all the subclass constructor, what do you think ?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionService.java [hunk 11]
```diff
@@ -87,15 +90,19 @@ protected void publish(Iterator<MetricValue> metrics) throws Exception {
     while (metrics.hasNext()) {
       // Encode each MetricRecord into bytes and make it an individual kafka message in a message set.
       MetricValue value = metrics.next();
-      recordWriter.encode(value, encoder);
-      // partitioning by the context
-      preparer.add(ByteBuffer.wrap(encoderOutputStream.toByteArray()), getPartitionKey(value));
-      encoderOutputStream.reset();
+      publishMetric(preparer, value);
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionService.java]: undo changes?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java [hunk 12]
```diff
@@ -83,7 +90,66 @@ public AggregatedMetricsEmitter load(EmitterKey key) throws Exception {
...
     LOG.trace("Start log collection for timestamp {}", timestamp);
-    Iterator<MetricValue> metricsItor = getMetrics(timestamp);
+
+    final Iterator<MetricValue> rawMetricsItor = getMetrics(timestamp);
+    Iterator<MetricValue> metricsItor;
+    if (!recordMetaMetrics) {
+      metricsItor = rawMetricsItor;
+    } else {
+      // wrap the raw metrics iterator with an iterator that will publish meta metrics
+      metricsItor = new AbstractIterator<MetricValue>() {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: Move this into a class/method to make the logic here simple and straightforward.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java [hunk 13]
```diff
@@ -83,7 +90,66 @@ public AggregatedMetricsEmitter load(EmitterKey key) throws Exception {
...
+                pendingMetrics.add(processDelayMetric);
+              }
+
+              return raw;
+            } else {
+              // publish pending metric
+              return pendingMetrics.poll();
+            }
+          } else {
+            if (!queuedAggregateMetrics) {
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: can we rename the variable to be more meaningful, ```flushTotalProcessedMetrics``` ?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java (hunk 12 above)
Comment (jwang47) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: Sounds good. I may also split this up into multiple iterator classes.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java (hunk 10 above)
Comment (jwang47) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: Sounds good.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java (hunk 13 above)
Comment (jwang47) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: I think `queuedAggregateMetrics` is meaningful: If true, that means we queued up the metrics representing the aggregate of the current metrics batch, and thus it's time to send out the queued up aggregate metrics.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java [hunk 14]
```diff
@@ -79,11 +84,79 @@ public AggregatedMetricsEmitter load(EmitterKey key) throws Exception {
...
+        protected MetricValue computeNext() {
+          if (rawMetricsItor.hasNext()) {
+
+            if (pendingMetrics.isEmpty()) {
+              // publish raw metric
+              MetricValue raw = rawMetricsItor.next();
+              numMetrics++;
+
+              // sample processing delay - assumes we mostly receive metrics in time order
+              if (raw.getTimestamp() - timestampLastSampled >= SAMPLING_DELAY_SEC) {
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionService.java]: should we also have another option to execute this (every 'n' seconds ?) , let's say we have ```SAMPLING_DELAY_SEC= 10s``` , and we emit 10k records at ```timestamp -1428023632``` and let's say it took 20seconds (hypothetically) to process, assuming we are continuing process at the same rate, by the time we reach ```timestamp -1428023732``` and execute the next iteration, the latency would be roughly close to 3 minutes, we might want to get this latency information more frequently right ?

**Comment by jwang47** [comment]
Discussed this in person with @gsps1 and @chtyim. We decided instead of emitting a metric for processing delay in the iterator, we'd compute running statistics about the processing delay and emit min, max, avg gauges for the processing delays for that batch.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionService.java [hunk 15]
```diff
@@ -75,7 +76,7 @@ protected void startUp() throws Exception {
   }
 
   @Override
-  protected void publish(Iterator<MetricValue> metrics) throws Exception {
+  protected void publish(MetricsIterator metrics) throws Exception {
```
Comment (chtyim) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionService.java]: Why the change in the collector end?

Comment (jwang47) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/collect/KafkaMetricsCollectionService.java]: Right, this isn't needed. Will change.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsIterator.java [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+      return endOfData();
+    }
+  }
+
+  public Iterator<MetricValue> getMetaMetrics(long timestampMs) {
+    long timestamp = TimeUnit.MILLISECONDS.toSeconds(timestampMs);
+
+    MetricValue delayAvg = new MetricValue(
+      ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, "system"),
+      "metrics.global.processed.delay.avg", timestamp, (long) stats.getAverage(), MetricType.GAUGE);
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsIterator.java]: you need to calculate the difference right ?
stats.getAverage() - timestamp ? as we are looking for latency ? -- same for other metrics below

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsIterator.java [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+    if (rawMetricsItor.hasNext()) {
+      MetricValue metric = rawMetricsItor.next();
+      stats.gauge(metric.getValue());
+      return metric;
+    } else {
+      return endOfData();
+    }
+  }
+
+  public Iterator<MetricValue> getMetaMetrics(long timestampMs) {
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsIterator.java]: nit: don't pass timestamp , as its just current time, you can get it inside.

File: cdap-watchdog/src/test/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionServiceTest.java [hunk 18]
```diff
@@ -53,7 +53,7 @@ public void testPublish() throws InterruptedException {
 
     AggregatedMetricsCollectionService service = new AggregatedMetricsCollectionService() {
       @Override
-      protected void publish(Iterator<MetricValue> metrics) {
+      protected void publish(Iterator<MetricValue> metrics) throws Exception {
```
Comment (gsps1) [inline cdap-watchdog/src/test/java/co/cask/cdap/metrics/collect/AggregatedMetricsCollectionServiceTest.java]: unwanted change ? remove?

**Comment by jwang47** [comment]
Tested and working in distributed cluster.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsCollectorIterator.java [hunk 19]
```diff
@@ -0,0 +1,84 @@
...
+      return Iterators.emptyIterator();
+    }
+
+    long currentTimeMs = System.currentTimeMillis();
+    long currentTimeSec = TimeUnit.MILLISECONDS.toSeconds(currentTimeMs);
+
+    MetricValue delayAvg = new MetricValue(
+      ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, "system"),
+      "metrics.collector.processed.delay.avg", currentTimeSec,
+      currentTimeMs - TimeUnit.SECONDS.toMillis(processDelayStats.getAverage()), MetricType.GAUGE);
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsCollectorIterator.java]: hey, @abaranau mentioned this has to be counter during our meeting. if we receive multiple iterators within a second , we would override.

**Comment by gsps1** [comment]
just one comment @jwang47 LGTM otherwise.

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsCollectorIterator.java (hunk 19 above)
Comment (jwang47) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsCollectorIterator.java]: Did you mean histogram? I thought we were going to implement histogram later?

**Comment by jwang47** [comment]
@gsps1 can you take another look?

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsCollectorIterator.java (hunk 19 above)
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsCollectorIterator.java]: no i don't mean histogram, if we process multiple batches within a second, then we will only have latency metrics for the last batch as we override the other batch data. we should use counter right?

**Comment by sreevatsanraman** [comment]
Made the change @gsps1 suggested. Running a build here: 
http://builds.cask.co/browse/CDAP-DUT1322-18

**Comment by sreevatsanraman** [comment]
One more time. http://builds.cask.co/browse/CDAP-DUT1322-19

File: cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsCollectorIterator.java [hunk 20]
```diff
@@ -0,0 +1,85 @@
...
+
+    long currentTimeMs = System.currentTimeMillis();
+    long currentTimeSec = TimeUnit.MILLISECONDS.toSeconds(currentTimeMs);
+
+    Map<String, String> tags = ImmutableMap.of(Constants.Metrics.Tag.NAMESPACE, "system",
+                                               Constants.Metrics.Tag.COMPONENT, "metrics.processor");
+
+    MetricValue delaySum = new MetricValue(
+      tags, "processed.delay.sum", currentTimeSec,
+      currentTimeMs - TimeUnit.SECONDS.toMillis(processDelayStats.getSum()), MetricType.COUNTER);
```
Comment (gsps1) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsCollectorIterator.java]: when we get the sum, its the sum of timestamp, to calculate the sum of latency, we need to do
``(currentTimems * processDelayStats.getCount) - TimeUnit.SECONDS.toMillis(processDelayStats.getSum())``

or pass the currentTimems to the getSum and get the sum of latency from there.

Comment (sreevatsanraman) [inline cdap-watchdog/src/main/java/co/cask/cdap/metrics/iterator/MetricsCollectorIterator.java]: Fixed.

**Comment by gsps1** [comment]
LGTM :+1:

**Comment by sreevatsanraman** [comment]
will merge once the build passes: http://builds.cask.co/browse/CDAP-DUT1322-21
